### ISSUE TITLE
Migrate  Beta Channel away from AMO

### ISSUE THREAD

**Issue by lidel** [ipfs-shipyard/ipfs-companion#385]
FYSA there will be a change in a way we publish beta versions, I'll handle it soon, this issue is just for informational purposes.

### Background

> Hello,
>
> You’re receiving this message because you’re an add-on developer registered on AMO (addons.mozilla.org) and you are using the beta versions feature in at least one of your add-ons. We want to inform you about major changes coming to this feature.
>
> The add-ons ecosystem is very complex, which makes managing AMO code significantly difficult. Newer developments have caused certain feature sets to overlap and become nearly redundant. **As a way to reduce complexity on AMO and leverage existing features, we’ll stop supporting beta versions for add-ons.**
>
> **As an alternative to beta versions, we recommend using the self-distributed versions feature instead.** You can create an self-hosted version branch for your beta versions that you can distribute on your own. Automatic updates would also need to be managed by you. You can find more information about signing and distributing self-hosted versions on MDN web docs.
>
> Once this change is deployed, beta versions will not appear on listing pages anymore. Current beta version users will update to the stable add-on update branch once a higher version of your add-on is available. **We plan to deploy this change on February 22, 2018.**
>
> Regards,
> 
> The Add-ons Team

### References

- https://www.ghacks.net/2018/02/16/ublock-origin-development-add-on-for-firefox/
- https://discourse.mozilla.org/t/how-to-migrate-beta-channel-users/25351/2?u=lidel
- https://discourse.mozilla.org/t/maintaining-your-own-development-channel/25188/1?u=lidel

**Comment by lidel** [comment]
Ideally we would have CI/CD so that latest version of master branch that pass all tests and linting is automatically signed and published to our own beta channel.

What does not seem to work for public extensions yet:

> Right now, the biggest problem to automate a developer-hosted beta channel is that the signing of the add-on package (required to distribute the extension) must be done manually on amo. The current add-on signing API fails if the extension is available publicly on amo, which is generally the case.
> 
> – https://discourse.mozilla.org/t/maintaining-your-own-development-channel/25188/2?u=lidel

What we can do, it to grab version built from `master` and sign it manually, then attach it to a post at [ipfs-companion/releases](https://github.com/ipfs/ipfs-companion/releases).  That way users of regular Firefox will be able to use it like a regular version that does not disappear between restarts. 

Getting automated updates in beta channel may require additional work, will get back to it later.

**Comment by lidel** [comment]
## The Plan: Setting Up The New Beta Channel

We want to make it easier for people to install and run development versions.
Mozilla provides some tools and APIs that enable us to self-host a beta version of our extension. 

Brain dump with required steps:

1. Create a git tag for a beta release
1. Update `add-on/manifest.json` and 
   - append `(Beta Channel)` at the end of `name` 
   - append `beta${BUILD_NUMBER}`  at the end of `version` (`BUILD_NUMBER` is an env variable exposed by jenkins)
   - add `update_url` property  to point at [`update.json`](https://developer.mozilla.org/en-US/Add-ons/Updates) somewhere
     - This URL needs to be stable, as browsers will poll it once a day for updates of beta channel
     (Final releases published to Mozilla and Chrome stores will not have `update_url`, as it is filled automatically upstream)
1. Build the `.zip`
1. Sign the `.zip` via [`web-ext sign`](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/web-ext_command_reference#web-ext_sign)
    - I will do this manually for now, as it requires personal API keys to AMO
1. Publish `.zip` somewhere 
    (eg. `/ipfs/` path at public gateway should work ok, we can also attach `.zip` to release notes at github)
1. Add it to `update.json`, publish it at `update_url`

## Open Questions

### OQ1: Where to host  `update.json` ?
As noted in (2) and (6) we need to self-host the `update.json` manifest somewhere. 
It needs to be a permalink that does not change between releases.
Github Pages should be enough for Beta Channel (we can switch to IPNS/DNSLINK later).

### OQ2: Should we automate the beta channel?

Initial idea I had is to write scripts for all the above operations and run everything via Jenkins as an optional final step that is executed only for a specific branch.
To keep things explicit, we could have a `beta` branch and Jenkins would build every time we merge from `master` to `beta`. We don't want to build from `master` directly, as some QA needs to be applied. 

It is possible to automate signing  (4) but we need to:
  - create separate user at AMO
    - generate API KEY and SECRET at https://addons.mozilla.org/en-US/developers/addon/api/key/
    - store secrets in Jenkins configuration somehow and use them as commandline parameters or inject as environment variables (`$WEB_EXT_API_KEY` `$WEB_EXT_API_SECRET`) – in a way that does not leak secrets to the public (eg. via logs)

## Current Approach

My plan is to (unless anyone suggests to do things differently):  

- **OQ1**: use Github Pages for now (TLS provided by Github)
- **OQ2**: do manual releases for now, smoke test procedure,  solve all issues and problems, when everything works smoothly, than we can start work towards automating it

**Comment by alanshaw** [comment]
@lidel I like the sound of this, is it just for Firefox or does the self-distributed versions API cover Chrome/others also?

Would it be simpler to just release a separate add on as "companion canary". The idea being that it is always the latest and greatest features and existing companion add on would always be the stable version.

Then we don't have to self host anything...

**Comment by lidel** [comment]
Browsers have different policies when it comes to beta and self-hosted extensions.

## Chrome

Self-hosting Chrome extension is a platform-dependent mess (see "[What are the supported deployment options for extensions after this change?](https://www.chromium.org/developers/extensions-deployment-faq#TOC-What-are-the-supported-deployment-options-for-extensions-after-this-change-)"). 

But based on my experience they don't really do any in-depth reviews at their store (only automated ones), so we can just publish a beta version  to Chrome Web Store and it should work fine.

## Firefox

Firefox is harder. Publishing Beta to AMO as a separate extension (with different extension Id)  does not solve the problem of human reviews at extension store.  That is why we have a backup plan of self-hosting.

That being said, your suggestion is valid in that we should try official route before investing time in setup of self-hosted version.  Perhaps AMO is more lenient for extension marked as Beta/experimental. 

Note to self: there should be a checkbox(?) to mark extension as "experimental":

> **What does it mean if an add-on is "experimental"?**
>     Experimental add-ons have been marked by their developers as not suitable for a wide audience. They may have bugs or not work properly. Use caution when installing experimental add-ons and uninstall the add-on immediately if you notice problems. Learn more about our review process 

It sounds like something we could enable for Beta, increasing likelihood of it not being removed by a human reviewer :+1: 

I'll  look into this.

**Comment by lidel** [comment]
Ok, uBlock already tried that, and based on comment from  it was rejected:

> Update: I've received feedback from AMO, asking me to instead create a self-hosted version of the beta version of uBlock Origin:
>
>>     Please submit your development builds as self-hosted versions to your original add-on.
>
> So essentially **publishing a dev build on AMO is being refused**. I will need to read more about self-hosted versions and how (and if) this can be done on GitHub.
>
>  _(Source:  https://github.com/gorhill/uBlock/releases/tag/1.15.11b0)_

**TL;DR** We should be able to publish Beta for Chrome as a separate listing in official web store ([others do just that](https://chrome.google.com/webstore/detail/tampermonkey-beta/gcalenpjmijncebpfijmoaglllgpjagf)), but for Firefox we need to make a self-hosted version.

**Comment by lidel** [comment]
## Versioned Dev Builds  :heavy_check_mark:

To make QA and tests easier for everyone, Jenkins will produce versioned artifacts. 
Details in PR #397

## Dev Channel for Chrome :heavy_check_mark: 

I managed to create Dev Channel for Chromium-based browsers as a separate, unlisted extension in Chrome Web Store. This means it won't show in search (which is what we want), but everyone with direct link can install it and will get updates automatically. 

I am not adding it to README yet, as I want to smoke-test it for a day or two, make sure it won't get removed in first 24h. But if anyone with Chrome wants to give it a try, here is the link: 

- https://chrome.google.com/webstore/detail/ipfs-companion-dev-build/hjoieblefckbooibpepigmacodalfndh :wrench: 

## Dev Channel for Firefox :hourglass_flowing_sand: 

As noted in my previous comment, we need to self-host it. So far I made sure versioned dev builds are compatible with [addons-linter](https://github.com/mozilla/addons-linter). This means we can just grab them from Jenkins. 
Missing pieces: signing and setup related to `update_url`.  Will work on this next.

**Comment by lidel** [comment]
## Beta Channel for Firefox :heavy_check_mark:

Hello! I created PoC Beta Channel for Firefox in #400. 

If anyone wants to give it a try before PR is merged:

- First Beta is [ipfs_companion_dev_build_ea5c2af-2.2.0.7583-an+fx.xpi](https://ipfs.io/ipfs/QmXQLH4ib98NLgfhveitd1uK1rGuaKiimKVmrjdg6G7Lp6/ipfs_companion_dev_build_ea5c2af-2.2.0.7583-an+fx.xpi)
- <a href="https://support.mozilla.org/en-US/kb/unable-install-add-ons-extensions-or-themes#w_you-are-asked-to-download-the-add-on-rather-than-installing-it">Read this if you are asked to download the add-on, rather than installing it</a>


Feedback related to Beta channels is welcome!
